File: outco/matrix/robotPaths.js
/* 
        col
row    [[0,0],  
        [0,0]] 
       
Accessing:
    matrix[row][col]
    matrix[x][y]
    matrix[i][j]

    Example 1:
 Input:  
 [[ 0, 0, 0, 0],
  [ 0, 0, 0, 0],
  [ 0, 0, 0, 0]]

 Output: 38


Example 2:

 Input:  
 [[ 0, 0, 0],
  [ 0, 0, 0]]

 Output: 4


Example 3:
 Input:  
 [[ 0, 0, 0, 0, 0, 0, 0, 0],
  [ 0, 0, 0, 0, 0, 0, 0, 0],
  [ 0, 0, 0, 0, 0, 0, 0, 0],
  [ 0, 0, 0, 0, 0, 0, 0, 0],
  [ 0, 0, 0, 0, 0, 0, 0, 0]]

 Output: 7110272

 */

    

    const robotPaths = (matrix) => {
        let result = 0;
        //base cases: 
            // oob
            // already visited
            // reach destination
        let totalColumns = matrix[0].length -1
        let totalRows = matrix.length  -1
        const traverse = (row, col) => {
            if (row < 0 || col < 0 || col > matrix[0].length - 1 || row > matrix.length - 1){
                // out of bounds
                return;
            }
            if (matrix[row][col] === 1){
                //already visitedd
                return;
            }
            if (row === totalRows && col === totalColumns){
                result++
                return;
            }
            //toggle for tracking visited 
            matrix[row][col] = 1;
            traverse(row+1,col)
            traverse(row-1,col)
            traverse(row, col-1)
            traverse(row, col+1)
            // toggle back
            matrix[row][col] = 0;
        }
        traverse(0,0)
        return result;
    }

console.log(robotPaths( [[ 0, 0, 0, 0],
    [ 0, 0, 0, 0],
    [ 0, 0, 0, 0]]
  )) // 38

console.log(robotPaths( [[ 0, 0, 0],
    [ 0, 0, 0]])) // 4

console.log(robotPaths([[ 0, 0, 0, 0, 0, 0, 0, 0],
    [ 0, 0, 0, 0, 0, 0, 0, 0],
    [ 0, 0, 0, 0, 0, 0, 0, 0],
    [ 0, 0, 0, 0, 0, 0, 0, 0],
    [ 0, 0, 0, 0, 0, 0, 0, 0]])) // 7110272

File: c++journey/dsaDomains/matrix/robotPaths.cpp

#include <string>
#include <vector>
#include <iostream>
#include <algorithm>
#include <unordered_set>
using namespace std;


/*
 *  Problem 1:  Robot Paths
 *
 *  Developed on Sept. 22, 2023.
 * 
 *  Prompt:   Given a matrix of zeroes, determine how many unique paths exist
 *            from the top left corner to the bottom right corner
 *
 *  Input:    An Array of Array of ints (matrix)
 *  Output:   int
 *
 *  Example:  matrix = [[0,0,0,0],
 *                      [0,0,0,0],
 *                      [0,0,0,0]]
 *
 *            robotPaths(matrix) = 38
 *
 *
 *            matrix = [[0,0,0],
 *                      [0,0,0]]
 *
 *            robotPaths(matrix) = 4
 *
 *  Note:     From any point, you can travel in the four cardinal directions
 *            (north, south, east, west). A path is valid as long as it travels
 *            from the top left corner to the bottom right corner, does not go
 *            off of the matrix, and does not travel back on itself
 */

 class Solution{
   public:

   void depthFirstSearch(int rowInt, int colInt, vector<vector<int> > &matrix, unordered_set<string> &vSet, int &outputInt){
    if (rowInt == matrix.size() - 1 && colInt == matrix[0].size() - 1){
      outputInt++;
      return;
    }
    // OOB
    if (rowInt < 0 || rowInt >= matrix.size() || colInt < 0 || colInt >= matrix[0].size()){
        return;
    }
    string keyStr = to_string(rowInt)+"_"+to_string(colInt);
    // already visited this 1
    if (vSet.find(keyStr) != vSet.end()){
        return;
    }
    
    vSet.insert(keyStr);
    depthFirstSearch(rowInt + 1, colInt, matrix, vSet, outputInt);
    depthFirstSearch(rowInt - 1, colInt, matrix, vSet, outputInt);
    depthFirstSearch(rowInt, colInt + 1, matrix, vSet, outputInt);
    depthFirstSearch(rowInt, colInt - 1, matrix, vSet, outputInt);
    vSet.erase(keyStr);
    }

    int robotPaths(vector<vector<int> > matrix){
      int outputInt = 0;
      unordered_set<string> visitedSet;
      depthFirstSearch(0, 0, matrix, visitedSet, outputInt);
      return outputInt;
    }
 };
 
 int main(){
    Solution solutionX;
    vector<vector<int> > inputVec1 = {{0,0,0}, {0,0,0}};
    vector<vector<int> > inputVec2 = {{0,0,0,0},{0,0,0,0},{0,0,0,0}};
    int outputInt1 = solutionX.robotPaths(inputVec1);
    cout << "Result 1: " << outputInt1 << ". Expect 4." << endl;
    int outputInt2 = solutionX.robotPaths(inputVec2);
    cout << "Result 2: " << outputInt2 << ". Expect 38. " << endl;
    return 0;
 }